cpp/src/arrow/compute/expression.cc

```diff
@@ -536,14 +566,21 @@ Result<Expression> BindNonRecursive(Expression::Call call, bool insert_implicit_
   std::vector<TypeHolder> types = GetTypes(call.arguments);
   ARROW_ASSIGN_OR_RAISE(call.function, GetFunction(call, exec_context));
 
-  // First try and bind exactly
-  Result<const Kernel*> maybe_exact_match = call.function->DispatchExact(types);
-  if (maybe_exact_match.ok()) {
-    call.kernel = *maybe_exact_match;
-  } else {
-    if (!insert_implicit_casts) {
-      return maybe_exact_match.status();
+  bool maybe_need_dispatch_best = IsNeedDispatchBest(types, call.function_name);
```

Comment:
Instead of putting a special case here, I'd prefer to see ResolveDecimalAdditionOrSubtractionOutput amended. I think it would be sufficient to make the change

```diff
diff --git a/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc b/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
index 44f5fea79..8aa14aa3f 100644
--- a/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
+++ b/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc
@@ -499,8 +499,9 @@ Result<TypeHolder> ResolveDecimalBinaryOperationOutput(
   DCHECK_EQ(left_type.id(), right_type.id());
 
   int32_t precision, scale;
-  std::tie(precision, scale) = getter(left_type.precision(), left_type.scale(),
-                                      right_type.precision(), right_type.scale());
+  ARROW_ASSIGN_OR_RAISE(std::tie(precision, scale),
+                        ToResult(getter(left_type.precision(), left_type.scale(),
+                                        right_type.precision(), right_type.scale())));
   ARROW_ASSIGN_OR_RAISE(auto type, DecimalType::Make(left_type.id(), precision, scale));
   return std::move(type);
 }
@@ -508,7 +509,12 @@ Result<TypeHolder> ResolveDecimalBinaryOperationOutput(
 Result<TypeHolder> ResolveDecimalAdditionOrSubtractionOutput(
     KernelContext*, const std::vector<TypeHolder>& types) {
   return ResolveDecimalBinaryOperationOutput(
-      types, [](int32_t p1, int32_t s1, int32_t p2, int32_t s2) {
+      types,
+      [](int32_t p1, int32_t s1, int32_t p2,
+         int32_t s2) -> Result<std::pair<int32_t, int32_t>> {
+        if (s1 != s2) {
+          return Status::Invalid("");
+        }
         DCHECK_EQ(s1, s2);
         const int32_t scale = s1;
         const int32_t precision = std::max(p1 - s1, p2 - s2) + scale + 1;
```

Reply:
Yes, i have thought the changes here, it's much clear.
But if you only change the code here, the Decimal type calculation logic still has no way to enter the correct DispatchBest path.

Reply:
Maybe we misunderstand the logic here.
1. The `ResolveDecimalxxx` functions called in `BindNonRecursive` at [call.kernel->signature->out_type().Resolve](https://github.com/apache/arrow/blob/19e874ff8c6d960a58ee91a9605e133d99fec3f9/cpp/src/arrow/compute/expression.cc#L591), it'a called after the `DispatchExact` and `DispatchBest`.
2. [DispatchExact](https://github.com/apache/arrow/pull/40223/files#diff-81305966f880c8e3a51f4df82e8c042b2946f162c5b2f099b2f328097c3dd4caL541) will go into  [Function::DispatchExact](https://github.com/apache/arrow/blob/19e874ff8c6d960a58ee91a9605e133d99fec3f9/cpp/src/arrow/compute/function.cc#L297) which will always return ok and also won't go into the `Resolvexxx` functions. And `ScalarFunction` didn't override the `DispatchExact`.

Reply:
Ah, I see- sorry for the confusion. In addition to the changes I've suggested, we need to ensure output type resolution (and kernel init) must succeed before the exact kernel is accepted. For that the following should be sufficient

```diff
diff --git a/cpp/src/arrow/compute/expression.cc b/cpp/src/arrow/compute/expression.cc
index 8c59ad1df..a42bac175 100644
--- a/cpp/src/arrow/compute/expression.cc
+++ b/cpp/src/arrow/compute/expression.cc
@@ -536,14 +536,39 @@ Result<Expression> BindNonRecursive(Expression::Call call, bool insert_implicit_
   std::vector<TypeHolder> types = GetTypes(call.arguments);
   ARROW_ASSIGN_OR_RAISE(call.function, GetFunction(call, exec_context));
 
+  auto FinishBind = [&] {
+    compute::KernelContext kernel_context(exec_context, call.kernel);
+
+    if (call.kernel->init) {
+      const FunctionOptions* options =
+          call.options ? call.options.get() : call.function->default_options();
+      ARROW_ASSIGN_OR_RAISE(
+          call.kernel_state,
+          call.kernel->init(&kernel_context, {call.kernel, types, options}));
+
+      kernel_context.SetState(call.kernel_state.get());
+    }
+
+    ARROW_ASSIGN_OR_RAISE(
+        call.type, call.kernel->signature->out_type().Resolve(&kernel_context, types));
+
+    return Status::OK();
+  };
+
   // First try and bind exactly
   Result<const Kernel*> maybe_exact_match = call.function->DispatchExact(types);
   if (maybe_exact_match.ok()) {
     call.kernel = *maybe_exact_match;
-  } else {
+
+    if (FinishBind().ok()) {
+      return Expression(std::move(call));
+    }
+  }
+  {
     if (!insert_implicit_casts) {
       return maybe_exact_match.status();
     }
+
     // If exact binding fails, and we are allowed to cast, then prefer casting literals
     // first.  Since DispatchBest generally prefers up-casting the best way to do this is
     // first down-cast the literals as much as possible
@@ -576,20 +601,7 @@ Result<Expression> BindNonRecursive(Expression::Call call, bool insert_implicit_
     }
   }
 
-  compute::KernelContext kernel_context(exec_context, call.kernel);
-  if (call.kernel->init) {
-    const FunctionOptions* options =
-        call.options ? call.options.get() : call.function->default_options();
-    ARROW_ASSIGN_OR_RAISE(
-        call.kernel_state,
-        call.kernel->init(&kernel_context, {call.kernel, types, options}));
-
-    kernel_context.SetState(call.kernel_state.get());
-  }
-
-  ARROW_ASSIGN_OR_RAISE(
-      call.type, call.kernel->signature->out_type().Resolve(&kernel_context, types));
-
+  RETURN_NOT_OK(FinishBind());
   return Expression(std::move(call));
 }
```

Reply:
Thanks for your suggestion. The changes should make sense!